### PR TITLE
Fix AnalyzeAction response serialization

### LINKED ISSUE #44078: ES 6.8.1/7.2.0 varying responses on /<index>/_analyze request in a two node cluster
Windows 10, JRE version 1.8.0_201

1. Download and unpack elasticsearch zip archive
2. Add a row to the configuration file:
`node.max_local_storage_nodes: 2`
3. Run `elasticsearch.bat` in two separate powershell windows
4. Create an index: `PUT /foo`
5. Analysis request (against index) of text containing *only punctuation* returns different responses.
* **Elasticsearch 6.8.1**

Request:
`GET /foo/_analyze
{
  "text": "."
}`
Response: `{
   "tokens": []
}` alternates with empty response `{}`

Request against index with `explain` option:
`GET /foo/_analyze
{
  "text": ".",
  "explain": true
}`
Normal response 
```
{
   "detail": {
      "custom_analyzer": false,
      "analyzer": {
         "name": "default",
         "tokens": []
      }
   }
}
```
alternates with error response
```
{
   "error": {
      "root_cause": [
         {
            "type": "null_pointer_exception",
            "reason": null
         }
      ],
      "type": "null_pointer_exception",
      "reason": null
   },
   "status": 500
}
```
Logs of error response:
```
[2019-07-08T17:47:36,979][WARN ][r.suppressed             ] [S6SaS48] path: /foo/_analyze, params: {index=foo}
java.lang.NullPointerException: null
	at org.elasticsearch.action.admin.indices.analyze.DetailAnalyzeResponse$AnalyzeTokenList.toXContentWithoutObject(DetailAnalyzeResponse.java:299) ~[elasticsearch-6.8.1.jar:6.8.1]
	at org.elasticsearch.action.admin.indices.analyze.DetailAnalyzeResponse.toXContent(DetailAnalyzeResponse.java:140) ~[elasticsearch-6.8.1.jar:6.8.1]
	at org.elasticsearch.action.admin.indices.analyze.AnalyzeResponse.toXContent(AnalyzeResponse.java:261) ~[elasticsearch-6.8.1.jar:6.8.1]
	at org.elasticsearch.rest.action.RestToXContentListener.buildResponse(RestToXContentListener.java:47) ~[elasticsearch-6.8.1.jar:6.8.1]
	at org.elasticsearch.rest.action.RestToXContentListener.buildResponse(RestToXContentListener.java:42) ~[elasticsearch-6.8.1.jar:6.8.1]
	at org.elasticsearch.rest.action.RestToXContentListener.buildResponse(RestToXContentListener.java:34) ~[elasticsearch-6.8.1.jar:6.8.1]
	at org.elasticsearch.rest.action.RestResponseListener.processResponse(RestResponseListener.java:37) ~[elasticsearch-6.8.1.jar:6.8.1]
	at org.elasticsearch.rest.action.RestActionListener.onResponse(RestActionListener.java:47) [elasticsearch-6.8.1.jar:6.8.1]
	at org.elasticsearch.action.support.TransportAction$1.onResponse(TransportAction.java:85) [elasticsearch-6.8.1.jar:6.8.1]
	at org.elasticsearch.action.support.TransportAction$1.onResponse(TransportAction.java:81) [elasticsearch-6.8.1.jar:6.8.1]
	at org.elasticsearch.action.support.single.shard.TransportSingleShardAction$AsyncSingleAction$2.handleResponse(TransportSingleShardAction.java:268) [elasticsearch-6.8.1.jar:6.8.1]
	at org.elasticsearch.action.support.single.shard.TransportSingleShardAction$AsyncSingleAction$2.handleResponse(TransportSingleShardAction.java:252) [elasticsearch-6.8.1.jar:6.8.1]
	at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleResponse(TransportService.java:1104) [elasticsearch-6.8.1.jar:6.8.1]
	at org.elasticsearch.transport.TcpTransport$1.doRun(TcpTransport.java:985) [elasticsearch-6.8.1.jar:6.8.1]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-6.8.1.jar:6.8.1]
	at org.elasticsearch.common.util.concurrent.EsExecutors$DirectExecutorService.execute(EsExecutors.java:193) [elasticsearch-6.8.1.jar:6.8.1]
	at org.elasticsearch.transport.TcpTransport.handleResponse(TcpTransport.java:977) [elasticsearch-6.8.1.jar:6.8.1]
	at org.elasticsearch.transport.TcpTransport.messageReceived(TcpTransport.java:952) [elasticsearch-6.8.1.jar:6.8.1]
	at org.elasticsearch.transport.TcpTransport.inboundMessage(TcpTransport.java:763) [elasticsearch-6.8.1.jar:6.8.1]
	at org.elasticsearch.transport.netty4.Netty4MessageChannelHandler.channelRead(Netty4MessageChannelHandler.java:53) [transport-netty4-client-6.8.1.jar:6.8.1]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323) [netty-codec-4.1.32.Final.jar:4.1.32.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:297) [netty-codec-4.1.32.Final.jar:4.1.32.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
	at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:241) [netty-handler-4.1.32.Final.jar:4.1.32.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1434) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:965) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:656) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:556) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:510) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:470) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:909) [netty-common-4.1.32.Final.jar:4.1.32.Final]
	at java.lang.Thread.run(Unknown Source) [?:1.8.0_201]
```

* **Elasticsearch 7.2.0**

Request against empty `foo` index:
`GET /foo/_analyze
{
  "text": "."
}`
Response: `{
   "tokens": []
}` alternates with `{}`
But request with `explain` option always returns normal response.

### PR DESCRIPTION
**Pull request by cbuescher**
Currently we loose information about whether a token list in an AnalyzeAction
response is null or an empty list, because we write a 0 value to the stream in
both cases and deserialize to a null value on the receiving side. This change
fixes this so we write an additional flag indicating whether the value is null
or not, followed by the size of the list and its content.

Closes #44078

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/admin/indices/analyze/AnalyzeAction.java [hunk 1]
```diff
@@ -292,22 +293,29 @@ public static Request fromXContent(XContentParser parser, String index) throws I
...
         }
 
         public Response(StreamInput in) throws IOException {
             super.readFrom(in);
-            int size = in.readVInt();
-            if (size > 0) {
-                tokens = new ArrayList<>(size);
-                for (int i = 0; i < size; i++) {
-                    tokens.add(new AnalyzeToken(in));
+            if (in.getVersion().onOrAfter(Version.V_8_0_0)) {
```
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/action/admin/indices/analyze/AnalyzeAction.java]: I'd like to backport this to 7_3_0 at least if possible. Not sure of its feasible to 7_2_1, might complicate the whole backport story a bit.

File: server/src/test/java/org/elasticsearch/indices/analyze/AnalyzeActionIT.java [hunk 2]
```diff
@@ -387,5 +389,17 @@ public void testAnalyzeNormalizedKeywordField() throws IOException {
...
 
+    /**
+     * Input text that doesn't produce tokens should return an empty token list
+     */
+    public void testZeroTokenAnalysis() throws IOException {
+        assertAcked(prepareCreate("test"));
+        ensureGreen("test");
+
+        AnalyzeAction.Response analyzeResponse = client().admin().indices().prepareAnalyze("test", ".").get();
+        System.out.println(analyzeResponse);
```
Comment (romseygeek) [inline server/src/test/java/org/elasticsearch/indices/analyze/AnalyzeActionIT.java]: Leftover?

File: client/rest-high-level/src/test/java/org/elasticsearch/client/indices/AnalyzeResponseTests.java [hunk 3]
```diff
@@ -19,25 +19,24 @@
...
 import org.elasticsearch.common.xcontent.XContentParser;
 
 import java.io.IOException;
-import java.util.ArrayList;
 import java.util.Arrays;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
 
+@Repeat(iterations = 100)
```
Comment (romseygeek) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/indices/AnalyzeResponseTests.java]: Do you want to keep this or was it just for debugging?

File: client/rest-high-level/src/test/java/org/elasticsearch/client/indices/AnalyzeResponseTests.java [hunk 4]
```diff
@@ -19,25 +19,24 @@
...
 public class AnalyzeResponseTests extends AbstractResponseTestCase<AnalyzeAction.Response, AnalyzeResponse> {
 
     @Override
     protected AnalyzeAction.Response createServerTestInstance() {
         int tokenCount = randomIntBetween(1, 30);
         AnalyzeAction.AnalyzeToken[] tokens = new AnalyzeAction.AnalyzeToken[tokenCount];
         for (int i = 0; i < tokenCount; i++) {
-            tokens[i] = randomToken();
+            tokens[i] = org.elasticsearch.action.admin.indices.analyze.AnalyzeResponseTests.randomToken();
```
Comment (romseygeek) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/indices/AnalyzeResponseTests.java]: Can you use a static import here?

File: server/src/test/java/org/elasticsearch/action/admin/indices/analyze/AnalyzeResponseTests.java [hunk 5]
```diff
@@ -19,20 +19,25 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 
 import static org.hamcrest.Matchers.equalTo;
 
-public class AnalyzeResponseTests extends ESTestCase {
+public class AnalyzeResponseTests extends AbstractWireSerializingTestCase<AnalyzeAction.Response> {
```
Comment (romseygeek) [inline server/src/test/java/org/elasticsearch/action/admin/indices/analyze/AnalyzeResponseTests.java]: 👍

**Review by romseygeek — APPROVED** [review]
LGTM, a couple of nits but no need for another go round

File: server/src/test/java/org/elasticsearch/indices/analyze/AnalyzeActionIT.java (hunk 2 above)
Comment (cbuescher) [reply server/src/test/java/org/elasticsearch/indices/analyze/AnalyzeActionIT.java]: Already gone.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/indices/AnalyzeResponseTests.java (hunk 3 above)
Comment (cbuescher) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/indices/AnalyzeResponseTests.java]: Also gone.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/indices/AnalyzeResponseTests.java (hunk 4 above)
Comment (cbuescher) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/indices/AnalyzeResponseTests.java]: Actually I had to move this the random token generation to RandomObjects in `:test:framework` so it is usable from both server and client side response testing.

**Pull request merged**